src/libertem/analysis/raw.py

```diff
@@ -38,36 +39,37 @@ class PickResultSet(AnalysisResultSet):
 
 
 class PickFrameAnalysis(BaseAnalysis):
+    TYPE = 'UDF'
     """
     Pick a single, complete frame from a dataset
     """
-    def get_job(self):
-        assert self.dataset.shape.nav.dims in (1, 2, 3), "can only handle 1D/2D/3D nav currently"
-        x, y, z = (
-            self.parameters.get('x'),
-            self.parameters.get('y'),
+
+    def get_origin(self):
+        dims = self.dataset.shape.nav.dims
+        if dims not in (1, 2, 3):
+            raise ValueError(
+                "can only handle 1D/2D/3D nav currently, received %s dimensions" % dims
+            )
+        zyx = np.array([
             self.parameters.get('z'),
-        )
+            self.parameters.get('y'),
+            self.parameters.get('x'),
+        ])
+        messages = {
+            1: "Need x, not y and not z to index 1D dataset, received z=%s, y=%s, x=%s",
+            2: "Need x, y and not z to index 2D dataset, received z=%s, y=%s, x=%s",
+            3: "Need x, y z to index 3D dataset, received z=%s, y=%s, x=%s",
+        }
+        keep = zyx[-dims:]
+        drop = zyx[:-dims]
 
-        if self.dataset.shape.nav.dims == 1:
-            if x is None:
-                raise ValueError("need x to index 1D nav datasets")
-            if y is not None or z is not None:
-                raise ValueError("y and z must not be specified for 1D nav dataset")
-            origin = (x,)
-        elif self.dataset.shape.nav.dims == 2:
-            if x is None or y is None:
-                raise ValueError("need x/y to index 2D nav datasets")
-            if z is not None:
-                raise ValueError("z must not be specified for 2D nav dataset")
-            origin = (y, x)
-        elif self.dataset.shape.nav.dims == 3:
-            if x is None or y is None or z is None:
-                raise ValueError("need x/y/z to index 3D nav datasets")
-            origin = (z, y, x)
-        else:
-            raise ValueError("cannot operate on datasets with more than 3 nav dims")
+        if (None in keep) or not all((d is None for d in drop)):
```

Comment:
Nitpick: no need for the extra parenthesis in `all`, something like `all(d is None for d in drop)` works, too.